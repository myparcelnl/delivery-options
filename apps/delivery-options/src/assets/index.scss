@tailwind base;
@tailwind components;
@tailwind utilities;

$dhl-color: #FFCB01;

$carriers: (
  bpost: #EF2637,
  dhl: $dhl-color,
  dhleuroplus: $dhl-color,
  dhlforyou: $dhl-color,
  dhlparcelconnect: $dhl-color,
  dpd: #DC0032,
  postnl: #ED8C00,
  ups: #300,
);
