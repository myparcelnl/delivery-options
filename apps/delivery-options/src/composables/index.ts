export * from './events';
export * from './useActiveCarriers';
export * from './useBreakpoints';
export * from './useCurrentPlatform';
export * from './useDateFormat';
export * from './useFeatures';
export * from './useFormatDistance';
export * from './useLanguage';
export * from './useMostEcoFriendly';
export * from './useOptionsGroupedByCarrier';
export * from './usePickupLocation';
export * from './usePickupLocationsMap';
export * from './useProvideElementWidth';
export * from './useResolvedCarrier';
export * from './useResolvedDeliveryDates';
export * from './useResolvedDeliveryMoments';
export * from './useResolvedDeliveryOptions';
export * from './useResolvedPickupLocations';
export * from './useSelectedDeliveryMoment';
export * from './useSelectedPickupLocation';
export * from './useSelectedValues';
export * from './useShipmentOptionsOptions';
export * from './useTimeRange';
