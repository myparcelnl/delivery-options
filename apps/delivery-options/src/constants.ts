import {ShipmentOptionName} from '@myparcel/constants';

export const FIELD_DELIVERY_MOMENT = 'deliveryMoment';

export const FIELD_DELIVERY_DATE = 'deliveryDate';

export const FIELD_SHIPMENT_OPTIONS = 'shipmentOptions';

export const SHOWN_SHIPMENT_OPTIONS = [ShipmentOptionName.Signature, ShipmentOptionName.OnlyRecipient] as const;
