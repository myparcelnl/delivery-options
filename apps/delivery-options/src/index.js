export * from './config'
export * from './data'
export * from './helpers'
export * from './services'
