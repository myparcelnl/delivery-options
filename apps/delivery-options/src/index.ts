export * from './components';
export * from './composables';
export * from './constants';
export * from './form';
export * from './stores';
export * from './types';
export * from './utils';
