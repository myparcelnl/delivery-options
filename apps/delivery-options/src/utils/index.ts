export * from './calculateCutoffTime';
export * from './calculateDropOffDays';
export * from './createCarrierMarkerIcon';
export * from './createGetDeliveryOptionsParameters';
export * from './createNextDate';
export * from './createUtcDate';
export * from './getConfigFromWindow';
export * from './getConfigPriceKey';
export * from './getDefaultAddress';
export * from './getDeliveryTypePrice';
export * from './getResolvedCarrier';
export * from './getResolvedDeliveryType';
export * from './getResolvedValue';
export * from './hasSlotContent';
export * from './language';
export * from './padStart';
export * from './parseJson';
export * from './showDeveloperInfo';
export * from './stringToDate';
export * from './toTime';
