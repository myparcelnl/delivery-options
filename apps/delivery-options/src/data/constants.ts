import {ShipmentOptionName} from '@myparcel/constants';

export const FIELD_HOME_OR_PICKUP = 'homeOrPickup';

export const HOME_OR_PICKUP_HOME = 'home';

export const FIELD_DELIVERY_MOMENT = 'deliveryMoment';

export const FIELD_DELIVERY_DATE = 'deliveryDate';

export const FIELD_SHIPMENT_OPTIONS = 'shipmentOptions';

export const HOME_OR_PICKUP_PICKUP = 'pickup';

export const FIELD_PICKUP_LOCATION = 'pickupLocation';

export const SHOWN_SHIPMENT_OPTIONS = [ShipmentOptionName.Signature, ShipmentOptionName.OnlyRecipient] as const;

export const MAP_MARKER_CLASS_ACTIVE = 'active';

export const MAP_MARKER_CLASS_PREFIX = 'mp__carrier-marker';

export const SHOWN_OPENING_HOURS = 3;

export const OUTPUT_EVENT_DEBOUNCE_DELAY = 50;
