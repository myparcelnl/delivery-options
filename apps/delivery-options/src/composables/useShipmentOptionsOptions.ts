import {computed, type ComputedRef} from 'vue';
import {get} from '@vueuse/core';
import {
  ONLY_RECIPIENT_TITLE,
  type SelectOption,
  SIGNATURE_TITLE,
  type SupportedShipmentOptionName,
} from '@myparcel-do/shared';
import {ShipmentOptionName} from '@myparcel/constants';
import {getConfigPriceKey, getResolvedValue} from '../utils';
import {SHOWN_SHIPMENT_OPTIONS} from '../data';
import {useSelectedDeliveryMoment} from './useSelectedDeliveryMoment';
import {useResolvedDeliveryOptions} from './useResolvedDeliveryOptions';
import {useResolvedCarrier} from './useResolvedCarrier';
import {useLanguage} from './useLanguage';

const TRANSLATION_MAP = Object.freeze({
  [ShipmentOptionName.Signature]: SIGNATURE_TITLE,
  [ShipmentOptionName.OnlyRecipient]: ONLY_RECIPIENT_TITLE,
}) satisfies Record<SupportedShipmentOptionName, string>;

export const useShipmentOptionsOptions = (): ComputedRef<SelectOption[]> => {
  const deliveryOptions = useResolvedDeliveryOptions();
  const deliveryMoment = useSelectedDeliveryMoment();
  const {translate} = useLanguage();

  return computed(() => {
    const hasNoDeliveryOptions = deliveryOptions.loading.value || !deliveryOptions.value.length;
    const resolvedCarrier = useResolvedCarrier(deliveryMoment.value?.carrier);

    if (hasNoDeliveryOptions || !resolvedCarrier.value) {
      return [];
    }

    return SHOWN_SHIPMENT_OPTIONS.filter((option) => {
      return get(get(resolvedCarrier)?.allowedShipmentOptions)?.has(option);
    }).map((name) => {
      const match = get(deliveryMoment.value)?.shipmentOptions?.find((option) => option.name === name);

      const hasOnlyOneOption = match?.schema.enum.length === 1;

      const priceKey = getConfigPriceKey(name);

      return {
        label: translate(TRANSLATION_MAP[name]),
        value: name,
        disabled: hasOnlyOneOption,
        selected: hasOnlyOneOption ? match?.schema.enum[0] : false,
        price: getResolvedValue(priceKey, resolvedCarrier.value?.identifier) ?? undefined,
      } satisfies SelectOption;
    });
  });
};
