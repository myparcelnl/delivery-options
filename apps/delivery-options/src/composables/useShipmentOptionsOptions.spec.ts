import {useCarrierSettings, useResolvedCarrier} from './index';
import {afterEach, beforeEach, describe, expect, it} from 'vitest';
import {createPinia, setActivePinia} from 'pinia';
import {flushPromises} from '@vue/test-utils';
import {CarrierName, type ShipmentOptionName} from '@myparcel/constants';
import {getResolvedCarrier} from '../utils';
import {mockSelectedDeliveryOptions} from '../__tests__/utils/mockSelectedDeliveryOptions';
import {mockDeliveryOptionsConfig} from '../__tests__';
import {CarrierSetting, getShipmentOptionConfigMap, KEY_CONFIG} from '../../../../libs/shared/src';
import {useShipmentOptionsOptions} from './useShipmentOptionsOptions';

describe('useShipmentOptionsOptions', () => {
  beforeEach(async () => {
    setActivePinia(createPinia());

    useResolvedCarrier(CarrierName.PostNl);
    await flushPromises();
  });

  afterEach(() => {
    getResolvedCarrier.clear();
    useCarrierSettings.clear();
  });

  it.each(Object.entries(getShipmentOptionConfigMap()))(
    'should filter %s by whether it is enabled via the config',
    (shipmentOption, configKey) => {
      mockDeliveryOptionsConfig({
        [KEY_CONFIG]: {
          [CarrierSetting.AllowSignature]: true,
          [CarrierSetting.AllowOnlyRecipient]: true,
          [configKey]: false,
        },
      });
      mockSelectedDeliveryOptions();

      const result = useShipmentOptionsOptions();

      expect(result).toHaveLength(Object.keys(getShipmentOptionConfigMap()).length - 1);
      expect(result).not.toContainEqual({
        label: shipmentOption,
        value: shipmentOption,
        disabled: false,
        selected: false,
      });
    },
  );

  it.each(Object.entries(getShipmentOptionConfigMap()))(
    'should set %s to disabled based on enum',
    (shipmentOption, configKey) => {
      mockDeliveryOptionsConfig({[KEY_CONFIG]: {[configKey]: true}});
      mockSelectedDeliveryOptions({
        deliveryMoment: {
          shipmentOptions: [
            {
              name: shipmentOption as ShipmentOptionName,
              schema: {type: 'boolean', enum: [false]},
            },
          ],
        },
      });

      const result = useShipmentOptionsOptions();

      const option = result.value.find((item) => item.value === shipmentOption);
      expect(option).toMatchObject({disabled: true, selected: false});
    },
  );

  it.each(Object.entries(getShipmentOptionConfigMap()))(
    'should set %s to selected and disabled based on enum',
    (shipmentOption, configKey) => {
      mockDeliveryOptionsConfig({[KEY_CONFIG]: {[configKey]: true}});
      mockSelectedDeliveryOptions({
        deliveryMoment: {
          shipmentOptions: [
            {
              name: shipmentOption as ShipmentOptionName,
              schema: {type: 'boolean', enum: [true]},
            },
          ],
        },
      });

      const result = useShipmentOptionsOptions();

      const option = result.value.find((item) => item.value === shipmentOption);
      expect(option).toMatchObject({disabled: true, selected: true});
    },
  );
});
