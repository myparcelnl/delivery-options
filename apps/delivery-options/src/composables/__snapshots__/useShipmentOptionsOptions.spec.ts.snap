// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useShipmentOptionsOptions > should show only the configured options for package type 'digital_stamp' with carrier 'dhlforyou' 1`] = `[]`;

exports[`useShipmentOptionsOptions > should show only the configured options for package type 'digital_stamp' with carrier 'postnl' 1`] = `[]`;

exports[`useShipmentOptionsOptions > should show only the configured options for package type 'mailbox' with carrier 'dhlforyou' 1`] = `[]`;

exports[`useShipmentOptionsOptions > should show only the configured options for package type 'mailbox' with carrier 'postnl' 1`] = `[]`;

exports[`useShipmentOptionsOptions > should show only the configured options for package type 'package' with carrier 'dhlforyou' 1`] = `
[
  {
    "disabled": true,
    "label": "signatureTitle",
    "price": 0,
    "selected": true,
    "value": "signature",
  },
  {
    "disabled": true,
    "label": "onlyRecipientTitle",
    "price": 0,
    "selected": true,
    "value": "only_recipient",
  },
]
`;

exports[`useShipmentOptionsOptions > should show only the configured options for package type 'package' with carrier 'postnl' 1`] = `
[
  {
    "disabled": true,
    "label": "signatureTitle",
    "price": 0,
    "selected": true,
    "value": "signature",
  },
  {
    "disabled": true,
    "label": "onlyRecipientTitle",
    "price": 0,
    "selected": true,
    "value": "only_recipient",
  },
]
`;

exports[`useShipmentOptionsOptions > should show only the configured options for package type 'package_small' with carrier 'dhlforyou' 1`] = `[]`;

exports[`useShipmentOptionsOptions > should show only the configured options for package type 'package_small' with carrier 'postnl' 1`] = `
[
  {
    "disabled": true,
    "label": "signatureTitle",
    "price": 0,
    "selected": true,
    "value": "signature",
  },
  {
    "disabled": true,
    "label": "onlyRecipientTitle",
    "price": 0,
    "selected": true,
    "value": "only_recipient",
  },
]
`;
